### PR TITLE
feat(l1): add libmdbx features 

### PR DESCRIPTION
**Pull request by fborello-lambda**
**Motivation**

`mdbx` cannot be compiled for the `rv32im` target (used in L2 prover) so this introduces a feature to disable it in those cases, as we need to import some `ethereum_rust` crates in the program.

This PR also adds (De)Serialization for types that we need to send to the zkVM program (which are (de)serialized in a binary format)

**Description**

- Bump up `revm` version
- Set `features` to compile limbmdbx conditionally 
- Avoid `revm's` `default-features` &rarr; https://docs.rs/crate/revm/14.0.3/features
- The `blst` feature from `revm` imported by `ethereum_rust-evm` was causing some issues, so it was disabled with `default-features = false`.
- `mdbx` is compiled conditionally based on the `libmdbx` feature.
- added CI job for building the prover zkVM program.
- adds (de)serialization for types that need to be sent to the zkVM.

### CODE REVIEW

File: crates/common/types/transaction.rs [hunk 1]
```diff
@@ -939,6 +940,172 @@ mod serde_impl {
         }
     }
 
+    impl<'de> Deserialize<'de> for Transaction {
```
Comment (fkrause98) [inline crates/common/types/transaction.rs:943]: Why do we need to implement deserialize manually? Do we need to deserialize a special format?

File: crates/common/types/transaction.rs [hunk 2]
```diff
@@ -939,6 +940,172 @@ mod serde_impl {
...
+                        .map_err(serde::de::Error::custom)?
+                        .trim_start_matches("0x"),
+                        16,
+                    )
+                    .map_err(serde::de::Error::custom)?
+                        != 0,
+                    signature_r: r,
+                    signature_s: s,
+                }),
+                TxType::EIP4844 => todo!("EIP4844 transaction deserialization"),
```
Comment (fkrause98) [inline crates/common/types/transaction.rs:1103]: Remove this todo, use an Error instead.

File: crates/vm/db.rs [hunk 3]
```diff
@@ -53,9 +53,9 @@ impl revm::Database for StoreWrapper {
             .unwrap_or_else(|| RevmU256::ZERO))
     }
 
-    fn block_hash(&mut self, number: RevmU256) -> Result<RevmB256, Self::Error> {
+    fn block_hash(&mut self, number: u64) -> Result<RevmB256, Self::Error> {
```
[the remark this answers is not in the document]
Comment (fborello-lambda) [reply crates/vm/db.rs:56]: because we are using version `14.0.3` and  now it needs a `u64`.

File: crates/common/types/transaction.rs (hunk 1 above)
Comment (fborello-lambda) [reply crates/common/types/transaction.rs:943]: we need `de/serialization` for the prover.

File: crates/vm/vm.rs [hunk 4]
```diff
@@ -485,16 +484,20 @@ fn tx_env_from_generic(tx: &GenericTransaction, basefee: u64) -> TxEnv {
         chain_id: tx.chain_id,
         access_list: tx
             .access_list
-            .iter()
-            .map(|entry| {
-                (
-                    RevmAddress(entry.address.0.into()),
-                    entry
-                        .storage_keys
+            .clone()
```
Comment (fkrause98) [inline crates/vm/vm.rs:487]: You don't need this clone, simply use iter here

File: crates/vm/Cargo.toml [hunk 5]
```diff
@@ -4,27 +4,36 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
-ethereum_rust-core.workspace = true
-ethereum_rust-storage.workspace = true
-revm = { version = "10.0.0", features = [
+ethereum_rust-core = { path = "../common", default-features = false }
```
[the remark this answers is not in the document]
Comment (fborello-lambda) [reply crates/vm/Cargo.toml:7]: If we use the workspace syntax, and we want to import a crate as a standalone package, we wouldn't be able to "`configure`" it with features since we would have like a "circular" dependecy.

For example(using `ethereun_rust-vm/Cargo.toml`):
```toml
# [...] 

[dependencies]
ethereum_rust-core = { path = "../common", default-features = false }
ethereum_rust-storage = { path = "../storage/store", default-features = false }

# [...] 

[features]
default = ["libmdbx", "c-kzg", "blst"]
l2 = []
c-kzg = ["revm/c-kzg"]
blst = ["revm/blst"]
libmdbx = ["ethereum_rust-storage/default", "ethereum_rust-core/libmdbx"]
```

The `workspace's` .toml sets the `libmdbx` feature for the `storage` crate. And if we use `ethereum_rust-storage` using the workspace syntax it compiles `libmdbx`.

File: crates/storage/trie/trie.rs [hunk 6]
```diff
@@ -507,9 +525,10 @@ mod test {
         );
     }
 
+    #[cfg(feature = "libmdbx")]
```
[the remark this answers is not in the document]
Comment (fborello-lambda) [reply crates/storage/trie/trie.rs:528]: Not necessary, the default feature is "libmdbx" for each crate making use of it. If we run the tests using `--no-default-features` it will not compile `libmdbx` and will run just `in_memory` tests.

File: crates/storage/trie/test_utils.rs [hunk 7]
```diff
@@ -1,31 +1,34 @@
...
+        (TestNodes) Vec<u8> => Vec<u8>
+    );
 
-/// Creates a new DB on a given path
-pub fn new_db_with_path<T: Table>(path: PathBuf) -> Arc<Database> {
-    let tables = [table_info!(T)].into_iter().collect();
-    Arc::new(Database::create(Some(path), &tables).expect("Failed creating db with path"))
-}
+    /// Creates a new DB on a given path
+    pub fn new_db_with_path<T: Table>(path: PathBuf) -> Arc<Database> {
```
[the remark this answers is not in the document]
Comment (fborello-lambda) [reply crates/storage/trie/test_utils.rs:16]: These functions were already implemented, they are used for testing. I can create an issue if it's needed.

File: crates/vm/vm.rs (hunk 4 above)
Comment (fborello-lambda) [reply crates/vm/vm.rs]: fixed

File: crates/common/types/transaction.rs (hunk 2 above)
Comment (fborello-lambda) [reply crates/common/types/transaction.rs]: fixed

File: crates/storage/trie/test_utils.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (fkrause98) [reply crates/storage/trie/test_utils.rs:16]: nvm, I just saw this is just test_utils, it's fine then.

File: crates/common/types/transaction.rs [hunk 8]
```diff
@@ -71,6 +71,7 @@ pub struct EIP1559Transaction {
     pub signature_s: U256,
 }
 
+// TODO/FIXME: We must implement a custom Deserialize
```
[the remark this answers is not in the document]
Comment (fborello-lambda) [reply crates/common/types/transaction.rs:74]: Done

File: crates/common/types/transaction.rs (hunk 1 above)
Comment (fborello-lambda) [reply crates/common/types/transaction.rs:943]: it has to be a custom Deserialization method because we have a custom Serialization method for all transactions

File: crates/vm/vm.rs [hunk 9]
```diff
@@ -465,6 +461,9 @@ fn tx_env(tx: &Transaction) -> TxEnv {
             .map(|hash| B256::from(hash.0))
             .collect(),
         max_fee_per_blob_gas,
+        // TODO revise
```
[the remark this answers is not in the document]
Comment (fborello-lambda) [reply crates/vm/vm.rs:464]: done

File: crates/common/types/transaction.rs (hunk 1 above)
Comment (xqft) [reply crates/common/types/transaction.rs:943]: We refactored this to derive Deserialize for Transaction

File: crates/storage/store/storage.rs [hunk 10]
```diff
@@ -70,14 +68,13 @@ impl AccountUpdate {
...
 
 impl Store {
-    pub fn new(path: &str, engine_type: EngineType) -> Result<Self, StoreError> {
+    pub fn new(_path: &str, engine_type: EngineType) -> Result<Self, StoreError> {
         info!("Starting storage engine ({engine_type:?})");
         let store = match engine_type {
             #[cfg(feature = "libmdbx")]
             EngineType::Libmdbx => Self {
-                engine: Arc::new(LibmdbxStore::new(path)?),
+                engine: Arc::new(LibmdbxStore::new(_path)?),
```
Comment (mpaulucci) [inline crates/storage/store/storage.rs:76]: why did you change to _ if you're using it

Comment (fborello-lambda) [reply crates/storage/store/storage.rs:76]: because the `path` is only being used if the `libmdbx` feature is set and clippy was throwing lint warning with the flag disabled,  but since we have it as default i will change it back

File: crates/storage/trie/node_hash.rs [hunk 11]
```diff
@@ -119,6 +121,8 @@ impl Encodable for NodeHash {
     }
 }
 
+#[cfg(feature = "libmdbx")]
+
```
Comment (ilitteri) [inline crates/storage/trie/node_hash.rs:125]:
```suggestion
```

File: crates/vm/vm.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ilitteri) [reply crates/vm/vm.rs:464]: Let's link the issue to the comment.

File: crates/vm/vm.rs [hunk 12]
```diff
@@ -504,6 +506,9 @@ fn tx_env_from_generic(tx: &GenericTransaction, basefee: u64) -> TxEnv {
             .map(|hash| B256::from(hash.0))
             .collect(),
         max_fee_per_blob_gas: tx.max_fee_per_blob_gas.map(RevmU256::from),
+        // TODO revise
+        // https://eips.ethereum.org/EIPS/eip-7702
```
Comment (ilitteri) [inline crates/vm/vm.rs:510]: Ditto.

**Review by ilitteri — COMMENTED** [review]
LGTM

File: crates/common/types/transaction.rs (hunk 1 above)
Comment (fkrause98) [reply crates/common/types/transaction.rs]: Ok, let's add some comments stating why we're doing this, so future readers have context on why we're doing it this way, IMO this intent is not really clear from the code alone, but feel free to correct me.

File: crates/vm/vm.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (fborello-lambda) [reply crates/vm/vm.rs:464]: https://github.com/lambdaclass/lambda_ethereum_rust/issues/883

File: crates/vm/vm.rs (hunk 12 above)
Comment (fborello-lambda) [reply crates/vm/vm.rs:510]: https://github.com/lambdaclass/lambda_ethereum_rust/issues/883

**Review by mpaulucci — APPROVED** [review]
